### PR TITLE
Add DT definitions for Wi-Fi device

### PR DESCRIPTION
**Pull request by rado17**
Fixes SHEL-875

### CODE REVIEW

File: drivers/wifi/nrf700x/src/fmac_main.c [hunk 1]
```diff
@@ -38,10 +38,12 @@
 
 #endif /* !CONFIG_NRF700X_RADIO_TEST */
 
+#define DT_DRV_COMPAT nrf_wlan0
 LOG_MODULE_DECLARE(wifi_nrf, CONFIG_WIFI_NRF700X_LOG_LEVEL);
 
 struct nrf_wifi_drv_priv_zep rpu_drv_priv_zep;
 
+#define NRF_NODE(n) DT_NODELABEL(nrf_wlan##n)
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/src/fmac_main.c:46]: `NRF_WLAN_NODE`  ?

File: drivers/wifi/nrf700x/src/fmac_main.c [hunk 2]
```diff
@@ -801,10 +802,9 @@ ETH_NET_DEVICE_INIT(wlan0, /* name - token */
 #endif /* !CONFIG_NRF700X_STA_MODE */
 		    CONFIG_WIFI_INIT_PRIORITY, /* prio */
 		    &wifi_offload_ops, /* api */
-		    CONFIG_NRF_WIFI_IFACE_MTU); /*mtu */
+		    CONFIG_NRF_WIFI_IFACE_MTU); /*mtu */;
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/src/fmac_main.c:805]: Why is this needed?

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/src/fmac_main.c (hunk 1 above)
Comment (rado17) [reply drivers/wifi/nrf700x/src/fmac_main.c]: Done

File: drivers/wifi/nrf700x/src/net_if.c [hunk 3]
```diff
@@ -577,8 +577,8 @@ int nrf_wifi_if_start_zep(const struct device *dev)
 	vif_info.state = NRF_WIFI_FMAC_IF_OP_STATE_UP;
 
 	memcpy(vif_info.ifacename,
-	       dev->name,
-	       strlen(dev->name));
+	       "wlan0",
```
Comment (krish2718) [inline drivers/wifi/nrf700x/src/net_if.c:580]: why is this now hardcoded?

File: boards/arm/nrf7002dk_nrf5340/nrf5340_cpuapp_common.dts [hunk 4]
```diff
@@ -150,6 +151,10 @@ arduino_i2c: &i2c1 {
 	pinctrl-0 = <&uart0_default>;
 	pinctrl-1 = <&uart0_sleep>;
 	pinctrl-names = "default", "sleep";
+	nrf_wlan0: nrf_wlan0 {
```
Comment (krish2718) [inline boards/arm/nrf7002dk_nrf5340/nrf5340_cpuapp_common.dts:154]: why is this under UART? Shouldn't this be a separate node? And the convention seems to be using `nordic`?

**Review by krish2718 — CHANGES_REQUESTED** [review]

**Review by sachinthegreen — APPROVED** [review]

File: drivers/wifi/nrf700x/src/net_if.c (hunk 3 above)
Comment (rado17) [reply drivers/wifi/nrf700x/src/net_if.c:580]: RPU identifies the dev name by wlan0.

Comment (jukkar) [reply drivers/wifi/nrf700x/src/net_if.c:580]: You can change the name of the network interface by calling
```
int net_if_set_name(struct net_if *iface, const char *buf);
```
if needed. Requires that `CONFIG_NET_INTERFACE_NAME` is enabled.
By default the first wifi interface is wlan0, but that can be changed if needed.

Comment (krish2718) [reply drivers/wifi/nrf700x/src/net_if.c:580]: > RPU identifies the dev name by wlan0.

This should be taken care in the FMAC as its specific to UMAC.

File: boards/arm/nrf7002dk_nrf5340/nrf5340_cpuapp_common.dts (hunk 4 above)
Comment (rado17) [reply boards/arm/nrf7002dk_nrf5340/nrf5340_cpuapp_common.dts]: Moved to the root node.. Vendor name convention is limited to compatible property [(ref).](https://docs.zephyrproject.org/2.7.5/guides/dts/intro.html#important-properties) .  `nrf_wlan0` should be fine for node label. But let me know if you want me to change this.

Comment (krish2718) [reply boards/arm/nrf7002dk_nrf5340/nrf5340_cpuapp_common.dts]: Yeah, using both `nrf` and `nordic` seems a bit weird, but I don't have strong opinions.

File: drivers/wifi/nrf700x/src/net_if.c (hunk 3 above)
Comment (rado17) [reply drivers/wifi/nrf700x/src/net_if.c]: Done

File: boards/arm/nrf7002dk_nrf5340/nrf5340_cpuapp_common.dts (hunk 4 above)
Comment (rado17) [reply boards/arm/nrf7002dk_nrf5340/nrf5340_cpuapp_common.dts]: Changed to `nordic` instead of using `nrf`

File: drivers/wifi/nrf700x/src/fmac_main.c [hunk 5]
```diff
@@ -789,8 +791,7 @@ static const struct zep_wpa_supp_dev_ops wpa_supp_ops = {
 
 
 #ifdef CONFIG_NET_L2_ETHERNET
-ETH_NET_DEVICE_INIT(wlan0, /* name - token */
-		    "wlan0", /* driver name - dev->name */
+ETH_NET_DEVICE_DT_DEFINE(NRF_WLAN_NODE(0),
```
[the remark this answers is not in the document]
Comment (rado17) [reply drivers/wifi/nrf700x/src/fmac_main.c]: Done

File: drivers/wifi/nrf700x/src/fmac_main.c [hunk 6]
```diff
@@ -803,8 +804,7 @@ ETH_NET_DEVICE_INIT(wlan0, /* name - token */
 		    &wifi_offload_ops, /* api */
 		    CONFIG_NRF_WIFI_IFACE_MTU); /*mtu */
 #else
-DEVICE_DEFINE(wlan0, /* name - token */
-	      "wlan0", /* driver name - dev->name */
+DEVICE_DT_DEFINE(NRF_WLAN_NODE(0),
```
[the remark this answers is not in the document]
Comment (rado17) [reply drivers/wifi/nrf700x/src/fmac_main.c]: Done

File: drivers/wifi/nrf700x/src/net_if.c [hunk 7]
```diff
@@ -577,8 +577,8 @@ int nrf_wifi_if_start_zep(const struct device *dev)
 	vif_info.state = NRF_WIFI_FMAC_IF_OP_STATE_UP;
 
 	memcpy(vif_info.ifacename,
-	       dev->name,
-	       strlen(dev->name));
+		   dev->name,
+		   strlen(dev->name));
```
Comment (krish2718) [inline drivers/wifi/nrf700x/src/net_if.c:581]: unnecessary whitespace change

File: modules/hostap/src/supp_main.c [hunk 8]
```diff
@@ -43,8 +43,8 @@ K_SEM_DEFINE(z_wpas_ready_sem, 0, 1);
 #include <l2_packet/l2_packet.h>
 
 /* Should match with the driver name */
-#define DEFAULT_IFACE_NAME "wlan0"
-#define IFACE_MATCHING_PREFIX "wlan"
+#define DEFAULT_IFACE_NAME "nordic_wlan0"
```
Comment (krish2718) [inline modules/hostap/src/supp_main.c:46]: How does the DTS entry map to iface name? I don't see anywhere we are setting the iface name actively?

File: drivers/wifi/nrf700x/src/net_if.c (hunk 7 above)
Comment (rado17) [reply drivers/wifi/nrf700x/src/net_if.c]: Removed.

File: modules/hostap/src/supp_main.c (hunk 8 above)
Comment (rado17) [reply modules/hostap/src/supp_main.c]: Created an entry in chosen node for `zephyr,wifi` which can be referred for fetching the device info.

File: modules/hostap/src/supp_main.c [hunk 9]
```diff
@@ -306,8 +302,15 @@ static void iface_event_handler(struct net_mgmt_event_callback *cb,
...
+	const struct device *wifi_dev = NULL;
+
+	wifi_dev =  DEVICE_DT_GET(DT_CHOSEN(zephyr_wifi));
+	if (!wifi_dev) {
+		wpa_printf(MSG_ERROR, "Device instance not found");
+		return;
+	}
 
-	if (strncmp(ifname, IFACE_MATCHING_PREFIX, sizeof(IFACE_MATCHING_PREFIX) - 1) != 0) {
+	if (strncmp(ifname, wifi_dev->name, sizeof(wifi_dev->name) - 1) != 0) {
```
Comment (krish2718) [inline modules/hostap/src/supp_main.c:313]: This won't work, as it needs to match multiple interfaces.

Comment (rado17) [reply modules/hostap/src/supp_main.c]: Introduced a Kconfig option for prefix string.

File: modules/hostap/src/supp_main.c [hunk 10]
```diff
@@ -307,7 +303,8 @@ static void iface_event_handler(struct net_mgmt_event_callback *cb,
 {
 	const char *ifname = iface->if_dev->dev->name;
 
-	if (strncmp(ifname, IFACE_MATCHING_PREFIX, sizeof(IFACE_MATCHING_PREFIX) - 1) != 0) {
+	if (strncmp(ifname, CONFIG_NRF700X_WLAN_IF_PREFIX,
+		    sizeof(CONFIG_NRF700X_WLAN_IF_PREFIX)) != 0) {
```
Comment (jukkar) [inline modules/hostap/src/supp_main.c:307]: This is big no-no, you are now introducing NRF specific checks into generic code. Beside zephyr upstream is already using interface name and not the device name to match the needed interface. It would be better if the downstream code would use the interface name too. If one knows network interface name, it is always possible to figure out what is the device, but it is not always possible to do the reverse (if VLAN is enabled).

**Review by jukkar — CHANGES_REQUESTED** [review]

File: modules/hostap/src/supp_main.c (hunk 10 above)
Comment (rado17) [reply modules/hostap/src/supp_main.c:307]: Moved the prefix definition to WPA supplicant module.

File: modules/hostap/src/supp_main.c [hunk 11]
```diff
@@ -42,10 +42,6 @@ LOG_MODULE_REGISTER(wpa_supplicant, LOG_LEVEL_DBG);
 K_SEM_DEFINE(z_wpas_ready_sem, 0, 1);
 #include <l2_packet/l2_packet.h>
 
-/* Should match with the driver name */
-#define DEFAULT_IFACE_NAME "wlan0"
```
Comment (krish2718) [inline modules/hostap/src/supp_main.c:46]: Please put this changes in a separate commit, a commit shouldn't span multiple modules, makes is tough to cherry pick or revert.

File: drivers/wifi/nrf700x/src/net_if.c [hunk 12]
```diff
@@ -606,10 +606,7 @@ int nrf_wifi_if_start_zep(const struct device *dev)
 	       sizeof(vif_info));
 
 	vif_info.state = NRF_WIFI_FMAC_IF_OP_STATE_UP;
-
-	memcpy(vif_info.ifacename,
```
Comment (krish2718) [inline drivers/wifi/nrf700x/src/net_if.c:610]: these changes are unrelated to commit log

File: drivers/wifi/nrf700x/src/net_if.c [hunk 13]
```diff
@@ -715,10 +712,7 @@ int nrf_wifi_if_stop_zep(const struct device *dev)
 	       sizeof(vif_info));
 
 	vif_info.state = NRF_WIFI_FMAC_IF_OP_STATE_DOWN;
-
-	memcpy(vif_info.ifacename,
-	       dev->name,
-	       strlen(dev->name));
+	vif_info.if_index = vif_ctx_zep->vif_idx;
```
Comment (krish2718) [inline drivers/wifi/nrf700x/src/net_if.c:715]: these changes are unrelated to commit log

File: drivers/wifi/nrf700x/src/wpa_supp_if.c [hunk 14]
```diff
@@ -1921,9 +1921,7 @@ static int nrf_wifi_vif_state_change(struct nrf_wifi_vif_ctx_zep *vif_ctx_zep,
 	vif_ctx = def_dev_ctx->vif_ctx[vif_ctx_zep->vif_idx];
 
 	vif_state_info.state = state;
-	memcpy(vif_state_info.ifacename,
-	       vif_ctx_zep->ifname,
-	       strlen(vif_ctx_zep->ifname));
+	vif_state_info.if_index = vif_ctx_zep->vif_idx;
```
Comment (krish2718) [inline drivers/wifi/nrf700x/src/wpa_supp_if.c:1924]: these changes are unrelated to commit log

**Review by krish2718 — CHANGES_REQUESTED** [review]

File: modules/hostap/src/supp_main.c (hunk 11 above)
Comment (rado17) [reply modules/hostap/src/supp_main.c:46]: Added a separate commit

File: drivers/wifi/nrf700x/src/net_if.c (hunk 13 above)
Comment (rado17) [reply drivers/wifi/nrf700x/src/net_if.c:715]: Included in a separate commit

File: drivers/wifi/nrf700x/src/wpa_supp_if.c (hunk 14 above)
Comment (rado17) [reply drivers/wifi/nrf700x/src/wpa_supp_if.c:1924]: Included in a separate commit

File: drivers/wifi/nrf700x/src/net_if.c (hunk 12 above)
Comment (rado17) [reply drivers/wifi/nrf700x/src/net_if.c:610]: Included in a separate commit

File: modules/hostap/src/supp_main.c [hunk 15]
```diff
@@ -340,12 +337,19 @@ static void wait_for_interface_up(const char *iface_name)
...
+		wpa_printf(MSG_ERROR, "Device instance not found");
+		return;
+	}
 
 	if (ifname == NULL) {
 		return;
 	}
 
-	if (strncmp(ifname, DEFAULT_IFACE_NAME, strlen(ifname)) != 0) {
+	if (strncmp(ifname, wifi_dev->name, strlen(ifname)) != 0) {
```
Comment (krish2718) [inline modules/hostap/src/supp_main.c:352]: Even though its seldom used, we need a multi-interface solution for WPA supplicant, esp. as this is all upstream code.
And as changing interface manager isn't a trivial task, IMHO, we should stick a Kconfig of allowed prefixes and then use that to select managed interfaces in CSV form, something like below. WDYT?

```
config WPA_SUPP_MGD_IFACES_PREFIX
       string "WPA supplicant managed interfaces prefix"
       default "wlan"

CONFIG_WPA_SUPP_MGD_IFACES_PREFIX="wlan,esp,nrf"
```

Comment (rado17) [reply modules/hostap/src/supp_main.c:352]: >CONFIG_WPA_SUPP_MGD_IFACES_PREFIX="wlan,esp,nrf"

Board Kconfig seems to be a good place to have this configuration. Would you agree?

File: modules/hostap/src/supp_main.c (hunk 10 above)
Comment (jukkar) [reply modules/hostap/src/supp_main.c:307]: Eventually the code in supp_main.c will be moved to upstream so we cannot have here NRF specific Kconfig symbols. So this cannot be done like this. Also wifi driver specific changes need to be separated to different commit than the hostap related ones.

**Review by jukkar — CHANGES_REQUESTED** [review]

File: modules/hostap/src/supp_main.c (hunk 10 above)
Comment (rado17) [reply modules/hostap/src/supp_main.c]: Done. Instead of NRF specific config name, using a WPA specific config. Separated the commits as well.

File: modules/hostap/src/supp_main.c (hunk 15 above)
Comment (rado17) [reply modules/hostap/src/supp_main.c:374]: Done.

File: modules/hostap/src/supp_main.c [hunk 16]
```diff
@@ -306,9 +306,32 @@ static void iface_event_handler(struct net_mgmt_event_callback *cb,
 							uint32_t mgmt_event, struct net_if *iface)
 {
 	const char *ifname = iface->if_dev->dev->name;
+	char *tmp_csv;
+	char *prefix;
+	char *mgd_if_prefix = CONFIG_WPA_SUPP_MGD_IFACES_PREFIX;
```
Comment (jukkar) [inline modules/hostap/src/supp_main.c:307]: We can do many things in built time with following changes.
Instead of using a pointer, we could do this (as the string is small the impact to stack is negligible)
```
char mgd_if_prefix[] = CONFIG_WPA_SUPP_MGD_IFACES_PREFIX;
```

File: modules/hostap/src/supp_main.c [hunk 17]
```diff
@@ -306,9 +306,32 @@ static void iface_event_handler(struct net_mgmt_event_callback *cb,
...
+	bool found = false;
+
+	if (strlen(mgd_if_prefix) > 0) {
```
Comment (jukkar) [inline modules/hostap/src/supp_main.c:310]: and here we can use sizeof
```
if (sizeof(mgd_if_prefix) > 0) {
```

File: modules/hostap/src/supp_main.c [hunk 18]
```diff
@@ -306,9 +306,32 @@ static void iface_event_handler(struct net_mgmt_event_callback *cb,
...
+		tmp_csv = os_zalloc(strlen(mgd_if_prefix));
```
Comment (jukkar) [inline modules/hostap/src/supp_main.c:311]: We do not need this heap allocation anymore, as we can manipulate the `mgd_if_prefix` variable as we want.

File: modules/hostap/src/supp_main.c [hunk 19]
```diff
@@ -306,9 +306,32 @@ static void iface_event_handler(struct net_mgmt_event_callback *cb,
...
+			wpa_printf(MSG_ERROR, "Failed to allocate memory for prefix string");
+			return;
+		}
+		strcpy(tmp_csv, mgd_if_prefix);
+		prefix = strtok(tmp_csv, ",");
 
-	if (strncmp(ifname, IFACE_MATCHING_PREFIX, sizeof(IFACE_MATCHING_PREFIX) - 1) != 0) {
-		return;
+		while (prefix) {
+			if (strncmp(ifname, prefix, strlen(prefix)) == 0) {
```
Comment (jukkar) [inline modules/hostap/src/supp_main.c:320]: We can either write '\0` where the `,` is or use proper len (as we know the start and stop of the prefix) when doing the compare.

File: modules/hostap/Kconfig [hunk 20]
```diff
@@ -60,6 +60,10 @@ choice LIBC_IMPLEMENTATION
 	default NEWLIB_LIBC
 endchoice
 
+config WPA_SUPP_MGD_IFACES_PREFIX
```
Comment (krish2718) [inline modules/hostap/Kconfig:63]: `IFACE_PREFIXES`

File: modules/hostap/Kconfig [hunk 21]
```diff
@@ -60,6 +60,10 @@ choice LIBC_IMPLEMENTATION
...
+	string "WPA supplicant managed interfaces prefix"
```
Comment (krish2718) [inline modules/hostap/Kconfig:64]:
```suggestion
	string "WPA supplicant managed interface prefixes"
```

File: modules/hostap/src/supp_main.c (hunk 19 above)
Comment (rado17) [reply modules/hostap/src/supp_main.c:320]: Will use `strtok` which returns a null-terminated string.

File: modules/hostap/Kconfig (hunk 21 above)
Comment (rado17) [reply modules/hostap/Kconfig]: Done

File: modules/hostap/Kconfig (hunk 20 above)
Comment (rado17) [reply modules/hostap/Kconfig]: Done

File: modules/hostap/src/supp_main.c (hunk 18 above)
Comment (rado17) [reply modules/hostap/src/supp_main.c]: Done

File: modules/hostap/src/supp_main.c (hunk 16 above)
Comment (rado17) [reply modules/hostap/src/supp_main.c]: Done.

File: modules/hostap/src/supp_main.c (hunk 17 above)
Comment (rado17) [reply modules/hostap/src/supp_main.c]: Done

**Review by VivekUppunda — APPROVED** [review]

**Review by krish2718 — APPROVED** [review]
